### PR TITLE
Create and lint for Promise helpers

### PR DESCRIPTION
**Pull request by jridgewell**
- Adds a `Deferred` class, which creates a `{ promise, resolve, reject }` struct.
- Adds `tryResolve`, which is analogous to the TC39 proposal `Promise.try`
  - /cc @aghassemi 

Closes https://github.com/ampproject/amphtml/issues/15108.

### CODE REVIEW

File: src/promise.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * @template T
+ */
+export class Deferred {
+  constructor() {
+    /**
+     * @const {!Promise<T>}
+     */
+    this.promise = new /*OK*/Promise((resolve, reject) => {
+      /**
+       * @const {function((TYPE|IThenable<TYPE>|Thenable)=)}
```
Comment (rsimha) [inline src/promise.js]: `TYPE` seems to be tripping up the closure compiler.

https://travis-ci.org/ampproject/amphtml/jobs/375151957#L2181

https://github.com/google/closure-compiler/wiki/Bad-Type-Annotation

**Comment by rsimha** [comment]
And don't forget to bump up the size here:

https://github.com/ampproject/amphtml/blob/2edf9c37d04beca16f950e6a9e87995c0918d745/build-system/tasks/bundle-size.js#L24

https://travis-ci.org/ampproject/amphtml/jobs/375151958#L666

**Comment by calebcordry** [comment]
Is there a reason you didn't want these to live in `utils/promises.js`?

**Comment by jridgewell** [comment]
I forgot that file existed...

**Pull request merged**